feat: welcome releasepost experiment (#1438)
updatecli/website: content/en/blog/2024/05/welcome_releasepost.md

In 2024, there is no excuse to skip that critical piece of information.

The sad news is that I now spend a significant amount of time searching for those release notes on platforms like GitHub...

---

_What if I could visualize Updatecli changelogs from [www.updatecli.io](https://www.updatecli.io/changelogs/updatecli)?_

_How many people read release notes?_

_Does it even make sense to spend effort maintaining those release notes?_

---

To try to answer those questions, I made a small experiment:

Reposting automatically Updatecli release notes to [updatecli.io](https://updatecli.io/changelogs/updatecli), in a way that would allow me to reuse this pipeline for others projects.

Considering [updatecli.io](https://www.updatecli.io) is powered by [HUGO](https://gohugo.io), all I need is a way to automate the synchronization of GitHub release pages with my website's Git repository.

Updatecli already works great to automatically push file updates to git repositories.

> Welcome Releasepost

Reviewer: ```suggestion
Welcome Releasepost.
```

## Releasepost

> Releasepost can be downloaded from [github.com/updatecli/releasepost](https://github.com/updatecli/releasepost/releases)

Reviewer: ```suggestion
Releasepost can be downloaded from [github.com/updatecli/releasepost](https://github.com/updatecli/releasepost/releases).
```

It is an experimental command line tool that fetches information from a remote location like GitHub releases and then generates files locally such as:

```bash
.
├── changelogs
│   ├── v0.9.0.md
│   ├── v0.9.1.md
│   ├── v0.9.2.md
│   ├── v0.9.3.md
│   └── v0.9.4.md
└── index.md
```

Where the directory `changelogs` contains one file per changelog, and `index.md` a list of links to the various changelogs file.

In theory, it's as simple as that...

> Let's see the Releasepost configuration used to generate [www.updatecli.io./changelogs](https://www.updatecli.io/changelogs/updatecli)

Reviewer: ```suggestion
Let's see the Releasepost configuration used to generate [www.updatecli.io./changelogs](https://www.updatecli.io/changelogs/updatecli):
```

<details><summary>releasepost.yaml</summary>

```yaml
changelogs:
  - kind: github
    dir: content/en/changelogs/updatecli
    formats:
      - extension: asciidoc
        frontmatters: |
          ---
          title: "{{ .Changelog.Name }}"
          date: {{ .Changelog.PublishedAt }}
          ---
        indexfilename: _index
         indexfrontmatters: |
           ---
          title: "Changelogs"
          ---
      - extension: json
        indexfilename: _index
    spec:
      owner: updatecli

>       repository: udash

Author: ```suggestion
      repository: updatecli
```

```

sourced from <a href="https://github.com/updatecli/website/blame/master/.releasepost.yaml">releasepost.yaml</a>
</details>

The most important settings are:

* `spec.owner` and `spec.repository` specify the GitHub repository to monitor for release notes.

> * `dir` specifies where to generate the files

Reviewer: ```suggestion
* `dir` specifies where to generate the files.
```

* `formats` specifies the kind of files we want to generate so in this case json and Asciidoctor.

This configuration relies on the environment variable `RELEASEPOST_GITHUB_TOKEN` to set the GitHub credentials used to interact with the API.

> [HUGO](https://gohugo.io) relies on specific behaviors like frontmatters to provide page metadata or an index file that differ from other tools like [Docusaurus](https://docusaurus.io/).

Author: ```suggestion
[HUGO](https://gohugo.io) relies on specific behaviors like frontmatters to provide page metadata or an index file name that differ from other tools like [Docusaurus](https://docusaurus.io/).
```

> I must admit, I was envisioning a simpler configuration that would only contains the GitHub repository to monitor, but as always, the Devil hides in the details.

Reviewer: ```suggestion
I must admit, I envisioned a simpler configuration that would only contain the GitHub repository to monitor, but as always, the Devil hides in the details.
```

NOTE: More customization is available such as version filtering or the credentials to use but let's keep that for later.

You can try releasepost by running:

```bash

> export RELEASEPOST_GITHUB_TOKEN=ghp_xxx

Author: ```suggestion
export RELEASEPOST_GITHUB_TOKEN=<insert here a GitHub personnal token>
```

releasepost --config releasepost.yaml
```

## Updatecli

Running `releasepost --config` creates the files locally but it's only halfway.

We want to automatically publish them to our git repository.

> The next step was to write a Updatecli policy that would run `releasepost` and take care of opening pull requests when needed, such as this [one](https://github.com/updatecli/website/pull/1428).

Reviewer: ```suggestion
The next step was to write an Updatecli policy that would run `releasepost` and take care of opening pull requests when needed, such as this [one](https://github.com/updatecli/website/pull/1428).
```

### Updatecli Policy

> On the Updatecli project, we are using this policy:

Author: ```suggestion
On the Updatecli project, we are using this Updatecli policy:
```

#### To visualize the pipeline of this policy:

<details><summary><code>updatecli manifest show ghcr.io/updatecli/policies/releasepost/releasepost:0.1.0</code></summary>

```bash
+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "/tmp/updatecli/store/6a8e68ae473a5dafa270650cd74fd296e61abaf4820daf4b50e0e21c8649b710/updatecli.d/default.tpl"

SCM repository retrieved: 0

++++++++++++++++++
+ AUTO DISCOVERY +
++++++++++++++++++

++++++++++++++++++++++++++++++++++
+ DOCS: SYNCHRONIZE RELEASE NOTE +
++++++++++++++++++++++++++++++++++

name: 'docs: synchronize release note'
pipelineid: releasepost
targets:
    default:
        name: synchronize release notes
        kind: shell
        spec:
            changedif:
                kind: exitcode
                spec:
                    failure: 2
                    success: 1
                    warning: 0
            command: releasepost --dry-run="$DRY_RUN" --config .releasepost.yaml
            environments:
                - name: GITHUB_TOKEN
                - name: RELEASEPOST_GITHUB_TOKEN
                - name: PATH
version: 0.77.0
```

</details>

#### To visualize what this pipeline would change

> <details><summary><code>updatecli diff ghcr.io/updatecli/policies/releasepost/releasepost:0.1.0</code></summary>

Author: ```suggestion
<details><summary><code>updatecli diff ghcr.io/updatecli/policies/releasepost/releasepost:0.1.0</code></summary>

Please note that to make the following output smaller, the releasepost configuration was modified to monitor the GitHub repositroy `updatecli/udash` which contains far less releases than `updatecli/updatecli`
```

    # Run at 12:00 every Thursday
    - cron: '0 12 * * 4'
jobs:
  updatecli:
    if: github.ref == 'refs/heads/master'
    runs-on: ubuntu-latest
    steps:
      - name: "Checkout"
        uses: "actions/checkout@v4"

      - name: "Install Updatecli"
        uses: "updatecli/updatecli-action@v2"

      # releasepost is required by the Updatecli
      #   * policy ghcr.io/updatecli/policies/releasepost/releasepost
      - name: "Install Releasepost"
        uses: "updatecli/releasepost-action@main"

      - name: "Run updatecli"
        run: "updatecli compose apply --file update-compose.weekly.yaml"
        env:
          GITHUB_ACTOR: ${{ github.actor }}
          ## Used by Updatecli to interact with GitHub api
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          ## Used by Releasepost to interact with GitHub api
          RELEASEPOST_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

</details>

## Conclusion

Here we are, to automate release note publishing to a documentation website.

We need three files:

1. A releasepost configuration that specifies the GitHub repository to monitor.
2. An update-compose.yaml file listing Updatecli policies to enforce.

> 3. A values file listing the GitHub repository to monitor for release notes and the GitHub repository to copy those release notes

Reviewer: ```suggestion
3. A values file listing the GitHub repository to monitor for release notes and the GitHub repository to copy those release notes.
```

4. A GitHub action workflow to execute Updatecli periodically.

To answer my initial questions.

_What if I could visualize Updatecli changelogs from [www.updatecli.io](https://www.updatecli.io/changelogs/updatecli)?_

> It's working great for a few months now, and I am not looking back.

Reviewer: ```suggestion
It's been working great for a few months now, and I am not looking back.
```

_How many people read release notes?_

Traffic show me that people/crawler do care.

_Does it even make sense to spend effort maintaining those release notes?_

I do think so, manually crafted release notes are by far the best ones but they are also the most time-consuming. It's difficult to maintain them over time for every project so automatically generated release notes are a good alternative.

> I wish more project consider release notes as a critical component of their releases.

Reviewer: ```suggestion
I wish more projects consider release notes as a critical component of their releases.
```

## Next steps

> This project was built with flexibility in mind, so we could definitely envisioned building plugins for NPM, Gitlab releases, etc.

Reviewer: ```suggestion
This project was built with flexibility in mind, so we could definitely envision building plugins for NPM, Gitlab releases, etc.
```

> In the near by future, I am considering documenting how we automate release note publishing on Docusaurus website leveraging the versioning feature.

Reviewer: ```suggestion
In the near future, I am considering documenting how we automate release notes publishing on Docusaurus website by leveraging the versioning feature.
```

In that scenario, we only need release notes associated to the versioned website.

Example on [fleet.rancher.io/0.9/changelogs](https://fleet.rancher.io/0.9/changelogs/)

> All of this don't replace the need to write good release notes, we just try to make them more visible.

Reviewer: ```suggestion
All of this doesn't replace the need to write good release notes, we just try to make them more visible.
```

          ---
      - extension: json
        indexfilename: _index
    spec:
      owner: updatecli
      repository: updatecli
```

sourced from <a href="https://github.com/updatecli/website/blame/master/.releasepost.yaml">releasepost.yaml</a>
</details>

The most important settings are:

* `spec.owner` and `spec.repository` specify the GitHub repository to monitor for release notes.
* `dir` specifies where to generate the files.
* `formats` specifies the kind of files we want to generate so in this case json and Asciidoctor.

This configuration relies on the environment variable `RELEASEPOST_GITHUB_TOKEN` to set the GitHub credentials used to interact with the API.

[HUGO](https://gohugo.io) relies on specific behaviors like front-matters to provide page metadata or an index file name that differs from other tools like [Docusaurus](https://docusaurus.io/).

I must admit, I envisioned a simpler configuration that would only contain the GitHub repository to monitor, but as always, the Devil hides in the details.

NOTE: More customization is available such as version filtering or the credentials to use but let's keep that for later.

You can try releasepost by running:

```bash

> export RELEASEPOST_GITHUB_TOKEN=<insert here a GitHub personnal token>

Author: ```suggestion
export RELEASEPOST_GITHUB_TOKEN=<insert here a GitHub personal token>
```
